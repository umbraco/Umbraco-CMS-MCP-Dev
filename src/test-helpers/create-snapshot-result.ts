import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
