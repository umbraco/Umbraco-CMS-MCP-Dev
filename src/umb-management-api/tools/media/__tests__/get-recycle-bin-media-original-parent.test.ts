import GetRecycleBinMediaOriginalParentTool from "../get/get-recycle-bin-media-original-parent.js";
import { MediaBuilder } from "./helpers/media-builder.js";
import { MediaTestHelper } from "./helpers/media-test-helper.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";
import { TemporaryFileBuilder } from "../../temporary-file/__tests__/helpers/temporary-file-builder.js";
import MoveMediaToRecycleBinTool from "../put/move-to-recycle-bin.js";

const TEST_MEDIA_NAME = "_Test Media Original Parent";
const TEST_PARENT_MEDIA_NAME = "_Test Parent Media";

describe("get-recycle-bin-media-original-parent", () => {
  let originalConsoleError: typeof console.error;
  let tempFileBuilder: TemporaryFileBuilder;

  beforeEach(async () => {
    originalConsoleError = console.error;
    console.error = jest.fn();

    tempFileBuilder = await new TemporaryFileBuilder()
      .withExampleFile()
      .create();
  });

  afterEach(async () => {
    // Clean up in parallel to speed up tests
    await Promise.all([
      MediaTestHelper.cleanup(TEST_MEDIA_NAME),
      MediaTestHelper.cleanup(TEST_PARENT_MEDIA_NAME)
    ]);
    console.error = originalConsoleError;
  }, 10000);

  it("should return original parent information for recycled media", async () => {
    // Create a parent media folder
    const parentMediaBuilder = await new MediaBuilder()
      .withName(TEST_PARENT_MEDIA_NAME)
      .withFolderMediaType()
      .create();

    // Create child media under the parent
    const childMediaBuilder = await new MediaBuilder()
      .withName(TEST_MEDIA_NAME)
      .withImageMediaType()
      .withImageValue(tempFileBuilder.getId())
      .withParent(parentMediaBuilder.getId())
      .create();

    // Move child media to recycle bin
    await MoveMediaToRecycleBinTool().handler(
      {
        id: childMediaBuilder.getId()
      },
      { signal: new AbortController().signal }
    );

    // Get original parent information
    const result = await GetRecycleBinMediaOriginalParentTool().handler(
      {
        id: childMediaBuilder.getId()
      },
      { signal: new AbortController().signal }
    );

    // Verify parent information is returned first to get the ID
    const parsed = JSON.parse(result.content[0].text as string);
    expect(parsed).toHaveProperty('id');
    // Verify the parent ID matches our created parent
    expect(parsed.id).toBe(parentMediaBuilder.getId());

    // Use createSnapshotResult with the specific ID to normalize
    const normalizedResult = createSnapshotResult(result, parsed.id);
    expect(normalizedResult).toMatchSnapshot();
  });

  it("should return null for media that was at root level", async () => {
    // Create media at root level (no parent)
    const mediaBuilder = await new MediaBuilder()
      .withName(TEST_MEDIA_NAME)
      .withImageMediaType()
      .withImageValue(tempFileBuilder.getId())
      .create();

    // Move media to recycle bin
    await MoveMediaToRecycleBinTool().handler(
      {
        id: mediaBuilder.getId()
      },
      { signal: new AbortController().signal }
    );

    // Get original parent information
    const result = await GetRecycleBinMediaOriginalParentTool().handler(
      {
        id: mediaBuilder.getId()
      },
      { signal: new AbortController().signal }
    );

    // For null responses, don't use createSnapshotResult as it can't handle null
    expect(result).toMatchSnapshot();

    // Should return null for root-level items
    const parsed = JSON.parse(result.content[0].text as string);
    expect(parsed).toBeNull();
  });

  it("should handle invalid media id", async () => {
    // Test with non-existent ID - this should return an error
    const result = await GetRecycleBinMediaOriginalParentTool().handler(
      {
        id: "00000000-0000-0000-0000-000000000000"
      },
      { signal: new AbortController().signal }
    );

    // For error responses, don't use createSnapshotResult as it expects JSON
    expect(result).toMatchSnapshot();

    // Verify it's an error response
    expect(result.content[0].text).toContain('Error');
  });
});