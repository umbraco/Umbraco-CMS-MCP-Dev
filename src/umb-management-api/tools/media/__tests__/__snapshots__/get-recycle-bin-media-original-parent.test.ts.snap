// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-recycle-bin-media-original-parent should handle invalid media id 1`] = `
{
  "content": [
    {
      "text": "Error using get-recycle-bin-media-original-parent:
{
  "message": "Request failed with status code 404",
  "response": {
    "type": "Error",
    "title": "The media item could not be found",
    "status": 404,
    "operationStatus": "NotFound"
  }
}",
      "type": "text",
    },
  ],
}
`;

exports[`get-recycle-bin-media-original-parent should return null for media that was at root level 1`] = `
{
  "content": [
    {
      "text": "null",
      "type": "text",
    },
  ],
}
`;

exports[`get-recycle-bin-media-original-parent should return original parent information for recycled media 1`] = `
{
  "content": [
    {
      "text": "{"id":"00000000-0000-0000-0000-000000000000"}",
      "type": "text",
    },
  ],
}
`;
